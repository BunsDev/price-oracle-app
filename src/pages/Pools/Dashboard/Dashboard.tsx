import { useEffect, useState } from 'react';
import styled from 'styled-components';
import { BigNumber, constants } from 'ethers';
import { isUndefined } from 'lodash';

import { Card, Loading, Typography, EthLabel, ValueInUSD, SPACING_1050 } from '~/components/shared';
import { useAppSelector } from '~/hooks';
import { LockManager } from '~/types';

const SCard = styled(Card)`
  grid-template-columns: repeat(2, auto);
  grid-template-rows: repeat(2, auto);
  justify-items: left;
  max-width: ${SPACING_1050};
  margin: 0 auto;
`;

const Title = styled.div`
  grid-column-start: span 2;
`;

const DashboardHeader = styled.div``;

const TokenAmount = styled.div`
  display: flex;
  flex-direction: column;
`;

export function Dashboard() {
  const [totalLockedInUsd, setTotalLockedInUsd] = useState<BigNumber>(constants.Zero);
  const [totalRewardsInUsd, setTotalRewardsInUsd] = useState<BigNumber>(constants.Zero);

  const lockManagers = useAppSelector((state) => state.lockManagers.elements);
  const { usdPerEth } = useAppSelector((state) => state.prices);
  const isLoading = !lockManagers && totalLockedInUsd.isZero;

  const getTotalLocked = (lockManagerList: LockManager[]) => {
    let locked = constants.Zero;
    lockManagerList.forEach((lockManager) => {
      if (lockManager.locked) {
        locked = locked.add(lockManager.locked);
      }
    });
    return locked;
  };

  const getTotalRewardsInUsd = (lockManagerList: LockManager[]) => {
    let rewardsInEth = constants.Zero;
    lockManagerList.forEach((lockManager) => {
      if (lockManager.rewards) {
        const tokenRewardInEth = constants.WeiPerEther.mul(lockManager.rewards.tokenReward).div(
          lockManager.rewards.tokenPerEth
        );
        rewardsInEth = rewardsInEth.add(lockManager.rewards.ethReward).add(tokenRewardInEth);
      }
    });
    return rewardsInEth.mul(usdPerEth).div(constants.WeiPerEther);
  };

  const hasRewards = lockManagers && Object.values(lockManagers).every((lm) => !isUndefined(lm.rewards));
  const totalUserLocked = hasRewards ? getTotalLocked(Object.values(lockManagers)) : '0';

  useEffect(() => {
    if (lockManagers && usdPerEth && hasRewards) {
      setTotalLockedInUsd(BigNumber.from(usdPerEth).mul(totalUserLocked).div(constants.WeiPerEther));
      setTotalRewardsInUsd(getTotalRewardsInUsd(Object.values(lockManagers)));
    }
  }, [lockManagers, usdPerEth]);

  return (
    <SCard>
      <Title>
        <Typography color='primary'>Dashboard</Typography>
      </Title>
      <DashboardHeader>
        <Typography color='primary'>Total ETH locked</Typography>
      </DashboardHeader>
      <DashboardHeader>
        <Typography color='primary'>Claimable rewards</Typography>
      </DashboardHeader>

      {isLoading && (
        <>
          <TokenAmount>
            <Loading />
          </TokenAmount>
          <TokenAmount>
            <Loading />
          </TokenAmount>
        </>
      )}

      {!isLoading && (
        <>
          <TokenAmount>
            <EthLabel value={totalUserLocked.toString()} />
            <ValueInUSD color='disabled' value={totalLockedInUsd.toString()} approximate={true} />
          </TokenAmount>
          <TokenAmount>
            {/* <PriceLabel value={totalPriceLocked} /> */}
            <ValueInUSD color='primary' value={totalRewardsInUsd.toString()} approximate={true} />
          </TokenAmount>
        </>
      )}
    </SCard>
  );
}

export default Dashboard;
