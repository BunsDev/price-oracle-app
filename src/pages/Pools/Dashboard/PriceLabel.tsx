import styled from 'styled-components';
import { BigNumber } from 'ethers';
// import BigNumber from 'bignumber.js';

import { TokenIcon, Typography, SPACING_8 } from '~/components/shared';
import { getConfig } from '~/config';
// import useHumanizeTokenValue from '../../hooks/useHumanizeTokenValue';

const Container = styled.div`
  display: grid;
  grid-template-columns: auto auto;
  grid-column-gap: ${SPACING_8};
`;

const ValueContainer = styled.div``;

const Suffix = styled(Typography).attrs({
  variant: 'small',
  color: 'disabled',
})`
  margin-left: 2px;
  margin-top: -8px;
`;

const Value = styled(Typography)``;

interface Props {
  value: string | BigNumber;
}

export const PriceLabel = ({ value }: Props) => {
  // const humanized = useHumanizeTokenValue(value);
  const { WETH_ADDRESS } = getConfig().ADDRESSES;
  const humanized = {
    value: 100,
    suffix: 'M',
  };

  return (
    <Container>
      <TokenIcon
        src={`https://raw.githubusercontent.com/trustwallet/assets/master/blockchains/ethereum/assets/${WETH_ADDRESS}/logo.png`}
      />
      <ValueContainer>
        <Value>{humanized.value}</Value>
        <Suffix>{humanized.suffix}</Suffix>
      </ValueContainer>
    </Container>
  );
};

export const TokenLabel = ({ value, address }: { value: string | BigNumber; address: string }) => {
  // const humanized = useHumanizeTokenValue(value);

  const humanized = {
    value: 100,
    suffix: 'M',
  };

  return (
    <Container>
      <TokenIcon
        src={`https://raw.githubusercontent.com/trustwallet/assets/master/blockchains/ethereum/assets/${address}/logo.png`}
      />
      <ValueContainer>
        <Value>{humanized.value}</Value>
        <Suffix>{humanized.suffix}</Suffix>
      </ValueContainer>
    </Container>
  );
};
