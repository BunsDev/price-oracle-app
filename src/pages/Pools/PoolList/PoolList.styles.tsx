import styled from 'styled-components';
import { Card } from '~/components/shared';
import { Typography } from '~/components/shared';
import { SPACING_8, SPACING_16, SPACING_32, SPACING_192, SPACING_128, SPACING_256 } from '~/components/shared';

export const SCard = styled(Card)`
  margin-top: ${SPACING_32};
  overflow-x: auto;
  max-width: 1280px;
  margin: 30px auto;
  justify-items: left;
`;

export const Title = styled(Typography).attrs({
  color: 'secondary',
})``;

export const Table = styled.div`
  display: grid;
  grid-row-gap: ${SPACING_8};
`;

export const Row = styled.div`
  display: grid;
  grid-template-rows: auto;
  grid-template-columns: ${() =>
    [SPACING_32(), SPACING_192(), SPACING_128(), SPACING_128(), SPACING_192(), SPACING_256(), SPACING_128()].join(' ')};
  justify-content: right;
  align-items: center;
  margin: 0 auto;
  justify-items: left;
`;

export const Header = styled(Row)`
  margin: ${SPACING_16} auto;
`;

export const ButtonContainer = styled.div`
  display: flex;
  justify-content: right;
`;

export const LoaderContainer = styled.div`
  display: flex;
  justify-content: center;
`;

export const PriceAmountContainer = styled.div`
  display: flex;
  align-items: center;
`;

export const Divider = styled.div`
  padding-left: ${SPACING_8};
  padding-right: ${SPACING_8};
`;
