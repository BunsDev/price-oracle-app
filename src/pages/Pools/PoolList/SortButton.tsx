import { Icon, Typography, FONT_SIZE_12 } from '~/components/shared';

export type SortType = 'name' | 'apy' | 'locked' | 'claimable' | 'fee';

const SortButton = ({ text, type }: { text: string; type: SortType }) => {
  return (
    <div /*  onClick={() => handleClickSort()} */>
      <Typography color='secondary'>{text}</Typography>
      <Icon name='chevron-down' size={FONT_SIZE_12()} /* rotate={orderAsc ? 180 : 0} color={themeTextSecondary} */ />
    </div>
  );
};

export default SortButton;
