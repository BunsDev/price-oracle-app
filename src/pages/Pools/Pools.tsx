import Dashboard from './Dashboard/Dashboard';
import PoolList from './PoolList/PoolList';

function Pools() {
  return (
    <>
      <Dashboard />
      <PoolList />
    </>
  );
}

export default Pools;
