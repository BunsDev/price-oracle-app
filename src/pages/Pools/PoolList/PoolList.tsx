import {
  ButtonContainer,
  SCard,
  Row,
  Table,
  Title,
  Header,
  LoaderContainer,
  PriceAmountContainer,
  Divider,
} from './PoolList.styles';
import { Typography } from '~/components/shared';
import { PrimaryButton, SecondaryButton } from '~/components/shared';
import SearchInput from '~/components/shared/SearchInput';
import { PriceLabel, TokenLabel } from '../Dashboard/PriceLabel';
import SortButton from './SortButton';
import PoolIcon from '~/components/shared/PoolIcon';
import Loading from '~/components/shared/Loading';

const PoolList = () => {
  // temporary
  const isLoading = false;
  const handleClickLock = () => null;
  const handleClickClaimRewards = () => null;

  const pool = {
    name: 'TUSD-WETH',
    apy: '11',
    address: '0x0000000000085d4780B73119b644AE5ecd22b376',
    fee: '2',
    locked: 'test',
    claimable: 'test',
  };
  const pools = [pool, pool];

  const WETH_ADDRESS = '0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2';
  return (
    <>
      <SCard>
        <Title>Pools</Title>
        <SearchInput onChange={(e) => console.log(e.target.value)} />

        {isLoading && (
          <LoaderContainer>
            <Loading />
          </LoaderContainer>
        )}

        {!isLoading && (
          <Table>
            <Header>
              <Typography />
              <SortButton text='Name' type='name' /* pools={pools} onPoolsChanged={setPools} */ />
              <SortButton text='APY' type='apy' /* pools={pools} onPoolsChanged={setPools} */ />
              <SortButton text='Fee' type='fee' /* pools={pools} onPoolsChanged={setPools} */ />
              <SortButton text='Locked' type='locked' /* pools={pools} onPoolsChanged={setPools} */ />
              <SortButton text='Claimable rewards' type='claimable' /* pools={pools} onPoolsChanged={setPools} */ />
              <Typography />
            </Header>

            {pools.map((p) => (
              <Row key={p.address}>
                <Typography>
                  <PoolIcon pool={p} />
                </Typography>
                <Typography>{p.name}</Typography>
                <Typography>{p.apy}%</Typography>
                <Typography>{p.fee}%</Typography>
                <PriceAmountContainer>
                  <PriceLabel value={p.locked!} />
                </PriceAmountContainer>
                <PriceAmountContainer>
                  <PriceLabel value={p.claimable!} />
                  <Divider>/</Divider>
                  <TokenLabel value={p.claimable!} address={pool.address} />
                </PriceAmountContainer>
                <ButtonContainer>
                  <PrimaryButton onClick={() => handleClickLock()}>Lock</PrimaryButton>
                  <SecondaryButton onClick={() => handleClickClaimRewards()}>Claim rewards</SecondaryButton>
                </ButtonContainer>
              </Row>
            ))}
          </Table>
        )}
      </SCard>
    </>
  );
};

export default PoolList;
