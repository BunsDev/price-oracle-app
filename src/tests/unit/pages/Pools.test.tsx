import { render } from '~/tests';

import Pools from '~/pages/Pools';
import Dashboard from '~/pages/Pools/Dashboard/Dashboard';
import PoolList from '~/pages/Pools/PoolList/PoolList';

describe('Pool Page', () => {
  it('Should render Pools Page', () => {
    const { asFragment, getByText } = render(<Pools />);

    expect(getByText(/dashboard/i)).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
});

describe('Testing Pool page initial state', () => {
  it('"Total ETH locked" and "Calimable rewards" should start loading', () => {
    const dashboard = render(<Dashboard />);
    const loaderIcon = dashboard.getAllByTestId('loader-icon');
    expect(loaderIcon.length).toEqual(2);
  });

  it('"Pools" should start loading', () => {
    const poolList = render(<PoolList />);
    const loaderIcon = poolList.getAllByTestId('loader-icon');
    expect(loaderIcon.length).toEqual(1);
  });
});
