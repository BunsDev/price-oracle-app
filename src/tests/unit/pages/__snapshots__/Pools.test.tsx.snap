// Vitest Snapshot v1

exports[`Pool Page > Should render Pools Page 1`] = `
<DocumentFragment>
  <div
    class="sc-ehvNnt cKzixw"
  >
    <div
      class="sc-iBYQkv sc-jrcTuL blHKRc epIsAC sc-ipEyDJ sc-cUEOzv hhOCqg hhKkbK"
      color="rgba(255,255,255, 0.87)"
    >
      <div
        class="sc-jOiSOi kKqVgi"
      >
        <div
          class="sc-iBYQkv sc-jrcTuL hNdzWR epIsAC"
          color="rgba(255,255,255, 0.87)"
        >
          Dashboard
        </div>
      </div>
      <div
        class="sc-dPWrhe"
      >
        <div
          class="sc-iBYQkv sc-jrcTuL cWFtuj epIsAC"
          color="rgba(255,255,255, 0.87)"
        >
          Total ETH locked
        </div>
      </div>
      <div
        class="sc-dPWrhe"
      >
        <div
          class="sc-iBYQkv sc-jrcTuL cWFtuj epIsAC"
          color="rgba(255,255,255, 0.87)"
        >
          Claimable rewards
        </div>
      </div>
      <div
        class="sc-bCfvAP ShKCg"
      >
        <div
          class="sc-fLcnxK"
        >
          <i
            class="sc-hBxehG sc-bBABsx ccFtCM jFMzlb picon-loader"
            data-testid="loader-icon"
            name="loader"
          />
        </div>
      </div>
      <div
        class="sc-bCfvAP ShKCg"
      >
        <div
          class="sc-fLcnxK"
        >
          <i
            class="sc-hBxehG sc-bBABsx ccFtCM jFMzlb picon-loader"
            data-testid="loader-icon"
            name="loader"
          />
        </div>
      </div>
    </div>
    <div
      class="sc-iBYQkv sc-jrcTuL blHKRc epIsAC sc-ipEyDJ sc-cOxWqc hhOCqg drcGsI"
      color="rgba(255,255,255, 0.87)"
    >
      <div
        class="sc-iBYQkv sc-jrcTuL hNdzWR epIsAC sc-AHaJN"
        color="rgba(255,255,255, 0.87)"
      >
        Pools
      </div>
      <div
        class="sc-hhOBVt hvTQJO"
      >
        <i
          class="sc-hBxehG sc-gYbzsP ccFtCM fuCfJr picon-search"
          name="search"
        />
        <input
          aria-label="Search"
          class="sc-hHTYSt sc-dmctIk sc-cCjUiG fvoNMM jgGokt fhnoNX"
        />
      </div>
      <div
        class="sc-lbVpMG fBCmoe"
      >
        <div
          class="sc-fLcnxK"
        >
          <i
            class="sc-hBxehG sc-bBABsx ccFtCM jFMzlb picon-loader"
            data-testid="loader-icon"
            name="loader"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
